From 348165b4d0fa3f53b04e1ce4218156b89d1fed84 Mon Sep 17 00:00:00 2001
From: Antony Abee Prakash XV <antonyabee.prakashxv@adlinktech.com>
Date: Tue, 7 Jul 2020 20:02:56 +0530
Subject: [PATCH 18/18] Add-dual-display-support

This patch will add support for dual display (LVDS + HDMI) support.
        - LVDS panel n101bge (single channel, lcdif driven) is used along with HDMI display

Signed-off-by: Antony Abee Prakash XV <antonyabee.prakashxv@adlinktech.com>
---
 arch/arm64/boot/dts/adlink/Makefile           |   3 +
 ...dlink-lec-imx8m-1gb-ddr3l-dual-display.dts |  38 ++++++
 ...dlink-lec-imx8m-2gb-ddr3l-dual-display.dts |  10 ++
 .../adlink/adlink-lec-imx8m-dual-display.dts  | 115 ++++++++++++++++++
 4 files changed, 166 insertions(+)
 create mode 100644 arch/arm64/boot/dts/adlink/adlink-lec-imx8m-1gb-ddr3l-dual-display.dts
 create mode 100644 arch/arm64/boot/dts/adlink/adlink-lec-imx8m-2gb-ddr3l-dual-display.dts
 create mode 100644 arch/arm64/boot/dts/adlink/adlink-lec-imx8m-dual-display.dts

diff --git a/arch/arm64/boot/dts/adlink/Makefile b/arch/arm64/boot/dts/adlink/Makefile
index 077d059a591a..7df23b733d31 100644
--- a/arch/arm64/boot/dts/adlink/Makefile
+++ b/arch/arm64/boot/dts/adlink/Makefile
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 
 dtb-$(CONFIG_ARCH_LEC_IMX8M) += adlink-lec-imx8m.dtb \
+				adlink-lec-imx8m-dual-display.dtb \
 				adlink-lec-imx8m-dcss-g133han01.dtb \
 				adlink-lec-imx8m-lcdif-g104x1l03.dtb \
 				adlink-lec-imx8m-lcdif-g185xw01.dtb \
@@ -9,6 +10,7 @@ dtb-$(CONFIG_ARCH_LEC_IMX8M) += adlink-lec-imx8m.dtb \
 				adlink-lec-imx8m-lcdif-n101bge.dtb \
 				adlink-lec-imx8m-usb-otg-master.dtb \
 				adlink-lec-imx8m-2gb-ddr3l.dtb \
+				adlink-lec-imx8m-2gb-ddr3l-dual-display.dtb \
 				adlink-lec-imx8m-2gb-ddr3l-dcss-g133han01.dtb \
 				adlink-lec-imx8m-2gb-ddr3l-lcdif-g104x1l03.dtb \
 				adlink-lec-imx8m-2gb-ddr3l-lcdif-g185xw01.dtb \
@@ -17,6 +19,7 @@ dtb-$(CONFIG_ARCH_LEC_IMX8M) += adlink-lec-imx8m.dtb \
 				adlink-lec-imx8m-2gb-ddr3l-lcdif-n101bge.dtb \
 				adlink-lec-imx8m-2gb-ddr3l-usb-otg-master.dtb \
 				adlink-lec-imx8m-1gb-ddr3l.dtb \
+				adlink-lec-imx8m-1gb-ddr3l-dual-display.dtb \
 				adlink-lec-imx8m-1gb-ddr3l-dcss-g133han01.dtb \
 				adlink-lec-imx8m-1gb-ddr3l-lcdif-g104x1l03.dtb \
 				adlink-lec-imx8m-1gb-ddr3l-lcdif-g185xw01.dtb \
diff --git a/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-1gb-ddr3l-dual-display.dts b/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-1gb-ddr3l-dual-display.dts
new file mode 100644
index 000000000000..112d3cfc4d16
--- /dev/null
+++ b/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-1gb-ddr3l-dual-display.dts
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright 2017 NXP
+ * Copyright (C) 2017-2018 Pengutronix, Lucas Stach <kernel@pengutronix.de>
+ */
+
+#include "adlink-lec-imx8m-dual-display.dts"
+
+/ {
+       cpus {
+               /delete-node/ cpu@2;
+               /delete-node/ cpu@3;
+       };
+
+       pmu {
+                interrupt-affinity = <&A53_0>, <&A53_1>;
+        };
+
+        thermal-zones {
+                cpu-thermal {
+
+                        cooling-maps {
+                                map0 {
+                                        cooling-device =
+                                                <&A53_0 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+                                                <&A53_1 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+                               };
+                       };
+               };
+       };
+
+       resmem: reserved-memory {
+               /* change cma size to 384MB for 1GB DDR3L SKU */
+               linux,cma {
+                       size = <0 0x18000000>;
+               };
+       };
+};
diff --git a/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-2gb-ddr3l-dual-display.dts b/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-2gb-ddr3l-dual-display.dts
new file mode 100644
index 000000000000..817585b95032
--- /dev/null
+++ b/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-2gb-ddr3l-dual-display.dts
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright 2017 NXP
+ * Copyright (C) 2017-2018 Pengutronix, Lucas Stach <kernel@pengutronix.de>
+ */
+
+#include "adlink-lec-imx8m-dual-display.dts"
+
+/delete-node/ &pgc_vpu;
+/delete-node/ &vpu;
diff --git a/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-dual-display.dts b/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-dual-display.dts
new file mode 100644
index 000000000000..e2102dda0184
--- /dev/null
+++ b/arch/arm64/boot/dts/adlink/adlink-lec-imx8m-dual-display.dts
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright 2020 NXP.
+ */
+
+#include "adlink-lec-imx8m.dts"
+
+/ {
+       backlight: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&pwm1 0 1000000 0>;
+                brightness-levels = < 0  1  2  3  4  5  6  7  8  9
+                                     10 11 12 13 14 15 16 17 18 19
+                                     20 21 22 23 24 25 26 27 28 29
+                                     30 31 32 33 34 35 36 37 38 39
+                                     40 41 42 43 44 45 46 47 48 49
+                                     50 51 52 53 54 55 56 57 58 59
+                                     60 61 62 63 64 65 66 67 68 69
+                                     70 71 72 73 74 75 76 77 78 79
+                                     80 81 82 83 84 85 86 87 88 89
+                                     90 91 92 93 94 95 96 97 98 99
+                                    100>;
+                default-brightness-level = <80>;
+    		status = "okay";
+       };
+};
+
+&lcdif {
+	status = "okay";
+	max-res = <1600>, <1200>;
+
+	port@0 {
+		lcdif_mipi_dsi: endpoint {
+			remote-endpoint = <&mipi_dsi_in>;
+		};
+	};
+};
+
+&mipi_dsi {
+	status = "okay";
+	fsl,clock-drop-level = <2>;
+
+	panel@0 {
+		compatible = "inno,n101bge";
+		reg = <0>;
+		pinctrl-0 = <&pinctrl_panel>;
+		pinctrl-names = "default";
+		backlight = <&backlight>;
+		bridge-i2c-bus = <&i2c4>;
+		bridge-info = <3>;
+		bridge-gpios = <&gpio5 5 GPIO_ACTIVE_HIGH>;
+ 		enable-gpios = <&gpio1 0 GPIO_ACTIVE_HIGH>;   
+		power-gpios = <&gpio5 3 GPIO_ACTIVE_HIGH>;
+
+		dsi-lanes = <4>;
+		width-mm = <223>;
+		height-mm = <126>;
+		port {
+			panel_in: endpoint {
+				remote-endpoint = <&mipi_dsi_out>;
+			};
+		};
+	};
+
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			mipi_dsi_in: endpoint {
+				remote-endpoint = <&lcdif_mipi_dsi>;
+			};
+		};
+
+		port@1 {
+			reg = <1>;
+			mipi_dsi_out: endpoint {
+				remote-endpoint = <&panel_in>;
+			};
+		};
+	};
+};
+
+&dphy {
+	status = "okay";
+};
+
+&iomuxc {
+               pinctrl_bridge: bridgegrp {
+                       fsl,pins = <
+                               MX8MQ_IOMUXC_SPDIF_EXT_CLK_GPIO5_IO5    0x16                    // LCD_BRI_EN   
+                       >;
+               };
+
+               pinctrl_panel: panelgrp {
+			fsl,pins = <
+				MX8MQ_IOMUXC_GPIO1_IO00_GPIO1_IO0       0x16                                    // LVDS_BKLT_EN
+				MX8MQ_IOMUXC_SPDIF_EXT_CLK_GPIO5_IO5    0x16                    		// LCD_BRI_EN
+				MX8MQ_IOMUXC_SPDIF_TX_GPIO5_IO3         0x16                                    // LVDS_VDD_EN
+			>;
+               };
+
+               pinctrl_pwm1: pwm1grp {
+                       fsl,pins = <
+                               MX8MQ_IOMUXC_GPIO1_IO01_PWM1_OUT                0x06                            // LVDS_BKLTCTL
+                       >;
+               };
+};
+
+&pwm1 {
+       pinctrl-names = "default";
+       pinctrl-0 = <&pinctrl_pwm1>;
+       status = "okay";
+};
-- 
2.17.1

